executor/batch_checker.go

```diff
@@ -0,0 +1,252 @@
+			}
+		}
+	}
+
+	for _, row := range rows {
+		var handleKey *keyValueWithDupInfo
+		uniqueKeys := make([]*keyValueWithDupInfo, 0, nUnique)
+		newRowValue, err := b.encodeNewRow(ctx, t, row)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		// Append record keys and errors.
+		if t.Meta().PKIsHandle {
+			handle := row[handleCol.Offset].GetInt64()
+			handleKey = &keyValueWithDupInfo{
+				newKeyValue: keyValue{
+					key:   t.RecordKey(handle),
+					value: newRowValue,
+				},
+				dupErr: kv.ErrKeyExists.FastGen("Duplicate entry '%d' for key 'PRIMARY'", handle),
+			}
+		}
+
+		// append unique keys and errors
+		for _, v := range t.WritableIndices() {
+			if !v.Meta().Unique {
+				continue
+			}
+			colVals, err1 := v.FetchValues(row, nil)
+			if err1 != nil {
+				return nil, errors.Trace(err1)
+			}
+			// Pass handle = 0 to GenIndexKey,
+			// due to we only care about distinct key.
+			key, distinct, err1 := v.GenIndexKey(ctx.GetSessionVars().StmtCtx,
+				colVals, 0, nil)
+			if err1 != nil {
+				return nil, errors.Trace(err1)
+			}
+			// Skip the non-distinct keys.
+			if !distinct {
+				continue
+			}
+			colValStr, err1 := types.DatumsToString(colVals)
+			if err1 != nil {
+				return nil, errors.Trace(err1)
+			}
+			uniqueKeys = append(uniqueKeys, &keyValueWithDupInfo{
+				newKeyValue: keyValue{
+					key: key,
+				},
+				dupErr: kv.ErrKeyExists.FastGen("Duplicate entry '%s' for key '%s'",
+					colValStr, v.Meta().Name),
+			})
+		}
+		toBeCheckRows = append(toBeCheckRows, toBeCheckRow{
+			rowValue:   newRowValue,
+			handleKey:  handleKey,
+			uniqueKeys: uniqueKeys,
+		})
+	}
+	return toBeCheckRows, nil
+}
+
+// batchGetInsertKeys uses batch-get to fetch all key-value pairs to be checked for ignore or duplicate key update.
+func (b *batchChecker) batchGetInsertKeys(ctx sessionctx.Context, t table.Table, newRows []types.DatumRow) ([]toBeCheckRow, map[string][]byte, error) {
+	// Get keys need to be checked.
+	toBeCheckRows, err := b.getKeysNeedCheck(ctx, t, newRows)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+
+	// Batch get values.
+	nKeys := 0
+	for _, r := range toBeCheckRows {
+		nKeys += len(r.uniqueKeys)
+	}
+	batchKeys := make([]kv.Key, 0, nKeys)
+	for _, r := range toBeCheckRows {
+		if r.handleKey != nil {
+			batchKeys = append(batchKeys, r.handleKey.newKeyValue.key)
+		}
+		for _, k := range r.uniqueKeys {
+			batchKeys = append(batchKeys, k.newKeyValue.key)
+		}
+	}
+	values, err := kv.BatchGetValues(ctx.Txn(), batchKeys)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	return toBeCheckRows, values, nil
+}
+
+// initDupOldRowValue initializes dupOldRowValues which contain the to-be-updated rows from storage.
+func (b *batchChecker) initDupOldRowValue(ctx sessionctx.Context, t table.Table, newRows []types.DatumRow) (err error) {
+	b.dupOldRowValues = make(map[string][]byte, len(newRows))
+	handles := make([]int64, 0, len(newRows))
+	for _, r := range b.toBeCheckRows {
```

Comment:
How about we loop the `toBeCheckedRows` two times.
The first time we only set `dupOldRowValues` for handle key.
The second time we collect unique key handles, then batch get those handles.
Then we can extract two methods, `initDupOldRowFromHandleKey` and `initDupOldRowFromUniqueKey`.